### PR TITLE
[INFRA] Support vendoring of dependency headers

### LINKED ISSUE #1521: can't run the v3.0.1 tests against an installed copy of seqan
## Platform

  - SeqAn version: 3.0.1
  - Operating system: Linux mrc-tux 5.3.0-26-generic #28-Ubuntu SMP Wed Dec 18 05:37:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
  - Compiler: g++ (Debian 9.2.1-23) 9.2.1 20200110

### How to repeat the problem

Build and install the libseqan3-dev version 3.0.1 Debian package
and run this script: https://salsa.debian.org/med-team/seqan3/blob/master/debian/tests/run-tests from the root of the source package

### Expected behaviour
Tests build and execute successfully

### Actual behaviour

```
[  7%] Building CXX object alignment/matrix/CMakeFiles/edit_distance_trace_matrix_full_test.dir/edit_distance_trace_matrix_full_test.cpp.o
cd /tmp/autopkgtest.ZZbmhm/autopkgtest_tmp/build_unit/alignment/matrix && /usr/bin/ccache /usr/bin/c++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/tmp/autopkgtest.ZZbmhm/autopkgtest_tmp/build_unit/vendor/googletest/googletest/include -I/tmp/autopkgtest.ZZbmhm/autopkgtest_tmp/unit/SEQAN3_CLONE_DIR-NOTFOUND/test/include -isystem /usr/include/seqan3/submodules/sdsl-lite/include  -O3 -DNDEBUG   -pedantic -Wall -Wextra -Werror -std=c++17 -fconcepts -pthread -o CMakeFiles/edit_distance_trace_matrix_full_test.dir/edit_distance_trace_matrix_full_test.cpp.o -c /tmp/autopkgtest.ZZbmhm/autopkgtest_tmp/unit/alignment/matrix/edit_distance_trace_matrix_full_test.cpp
/tmp/autopkgtest.ZZbmhm/autopkgtest_tmp/unit/alignment/matrix/edit_distance_trace_matrix_full_test.cpp:11:10: fatal error: seqan3/test/pretty_printing.hpp: No such file or directory
   11 | #include <seqan3/test/pretty_printing.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

### PR DESCRIPTION
**Pull request by mr-c**
Fixes #1521

### CODE REVIEW

**Comment by marehr** [comment]
@mr-c Somehow the tests broke

File: build_system/seqan3-config.cmake [hunk 1]
```diff
@@ -168,6 +169,22 @@ if (NOT SEQAN3_INCLUDE_DIR)
...
+    if (EXISTS "${SEQAN3_INCLUDE_DIR}/seqan3/submodules")
+        file (GLOB submodules ${SEQAN3_INCLUDE_DIR}/seqan3/submodules/*/include)
+        foreach (submodule ${submodules})
+            if (IS_DIRECTORY ${submodule})
+                seqan3_config_print ("  …adding submodule include:  ${submodule}")
+                set (SEQAN3_DEPENDENCY_INCLUDE_DIRS ${submodule} ${SEQAN3_DEPENDENCY_INCLUDE_DIRS})
+            endif ()
+        endforeach ()
+    endif ()
+endif ()
```
Comment (marehr) [inline build_system/seqan3-config.cmake]: Can you shortly explain how the folder structure looks like at debian?

Is it 

```
PATH/seqan3/version.hpp
PATH/seqan3/alignment/*
PATH/seqan3/core/*
...
PATH/seqan3/submodules/
```
So basically the content of `seqan3/include/seqan3/` merged with `seqan3/submodules`?

File: test/seqan3-test.cmake [hunk 2]
```diff
@@ -40,7 +40,12 @@ file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
 add_library (seqan3_test INTERFACE)
 target_compile_options (seqan3_test INTERFACE "-pedantic"  "-Wall" "-Wextra" "-Werror" "-DNO_WARN_X86_INTRINSICS")
 target_link_libraries (seqan3_test INTERFACE "seqan3::seqan3" "pthread")
-target_include_directories (seqan3_test INTERFACE "${SEQAN3_CLONE_DIR}/test/include/")
+if (SEQAN3_CLONE_DIR)
+    target_include_directories (seqan3_test INTERFACE "${SEQAN3_CLONE_DIR}/test/include/")
+else ()
+    target_include_directories (seqan3_test INTERFACE "${CMAKE_CURRENT_LIST_DIR}/include/")
+endif ()
```
Comment (marehr) [inline test/seqan3-test.cmake]: If `SEQAN3_CLONE_DIR/test == CMAKE_CURRENT_LIST_DIR` the second version would suffice.

**Comment by marehr** [comment]
I will add this PR as discussion for our next meeting on monday. 

* SEQAN_LICENSE_DIR This seems only test related, so we might want to name it differently.

File: build_system/seqan3-config.cmake (hunk 1 above)
Comment (eseiler) [reply build_system/seqan3-config.cmake]: The current install script on Debian is
```
include/seqan3 /usr/include/ 
build_system/seqan3-config.cmake /usr/share/cmake/
# range-v3 can be removed once https://github.com/seqan/seqan3/pull/1267 is merged and released
submodules/range-v3/include/range /usr/include/seqan3/submodules/range-v3/include/
submodules/range-v3/include/meta /usr/include/seqan3/submodules/range-v3/include/
submodules/range-v3/include/concepts /usr/include/seqan3/submodules/range-v3/include/
submodules/range-v3/include/std /usr/include/seqan3/submodules/range-v3/include/
submodules/sdsl-lite/include/sdsl /usr/include/seqan3/submodules/sdsl/include/
```

So it is as you assumed, i.e.
```
/usr/include/seqan3/version.hpp
/usr/include/seqan3/alignment/*
/usr/include/seqan3/submodules/rang-v3/*
```

Comment (h-2) [reply build_system/seqan3-config.cmake]: The submodules really should be their own packages!

Comment (mr-c) [reply build_system/seqan3-config.cmake]: @h-2 Well, now that I have packaged a snapshot of the development tree I'll see if we can use the Debian range-v3 package.

As you know, we are all waiting on `sdsl-light` to make a release :crying_cat_face:

Comment (marehr) [reply build_system/seqan3-config.cmake]: @eseiler, @mr-c Can you provide the link to the scripts?

Comment (mr-c) [reply build_system/seqan3-config.cmake]: @marehr https://salsa.debian.org/med-team/seqan3/tree/master/debian

Comment (marehr) [reply build_system/seqan3-config.cmake]: Thank you! I replaced this part with the method in #1429.

Comment (marehr) [reply build_system/seqan3-config.cmake]: This can now be removed.

File: test/seqan3-test.cmake (hunk 2 above)
Comment (marehr) [reply test/seqan3-test.cmake]: Do you still know why we needed this? Is the file structure in debian differently?

File: test/unit/argument_parser/detail/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_INCLUDE_DIR}")
+add_definitions(-DSEQAN_LICENSE_DIR="${seqan3_test_unit_BINARY_DIR}/..")
```
Comment (marehr) [inline test/unit/argument_parser/detail/CMakeLists.txt]: This should be renamed to `SEQAN3_TEST_LICENSE_DIR`

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 4]
```diff
@@ -257,7 +257,7 @@ TEST(help_page_printing, copyright)
     const char * argvCopyright[] = {"./copyright", "--copyright"};
     argument_parser copyright("myApp", 2, argvCopyright);
 
-    std::ifstream license_file{std::string{{SEQAN_INCLUDE_DIR}} + "/../LICENSE.md"};
+    std::ifstream license_file{std::string{{SEQAN_LICENSE_DIR}} + "/LICENSE.md"};
```
Comment (marehr) [inline test/unit/argument_parser/detail/format_help_test.cpp]: SEQAN3_TEST_LICENSE_DIR

File: test/seqan3-test.cmake (hunk 2 above)
Comment (mr-c) [reply test/seqan3-test.cmake]: Yes, In Debian we install the tests for running afterwards

**Comment by mr-c** [comment]
Any hint as to why `format_help_test` broke?

**Comment by marehr** [comment]
> Any hint as to why `format_help_test` broke?

Not sure, I have to investigate

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 5]
```diff
@@ -275,7 +275,7 @@ TEST(help_page_printing, copyright)
     const char * argvCopyright[] = {"./copyright", "--copyright"};
     argument_parser copyright("myApp", 2, argvCopyright);
 
-    std::ifstream license_file{std::string{{SEQAN_INCLUDE_DIR}} + "/../LICENSE.md"};
+    std::ifstream license_file{std::string{{SEQAN3_TEST_LICENSE_DIR}} + "/LICENSE.md"};
```
Comment (marehr) [inline test/unit/argument_parser/detail/format_help_test.cpp]: He can't find this one.

File: test/unit/argument_parser/detail/CMakeLists.txt [hunk 6]
```diff
@@ -1,4 +1,4 @@
-add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_INCLUDE_DIR}")
+add_definitions(-DSEQAN3_TEST_LICENSE_DIR="${seqan3_test_unit_BINARY_DIR}/..")
```
Comment (marehr) [inline test/unit/argument_parser/detail/CMakeLists.txt]: That means this path is somehow wrong

Comment (marehr) [reply test/unit/argument_parser/detail/CMakeLists.txt]: Can you try this version?

```patch
diff --git a/test/unit/argument_parser/detail/CMakeLists.txt b/test/unit/argument_parser/detail/CMakeLists.txt
index 6f116f295..060a8dcb8 100644
--- a/test/unit/argument_parser/detail/CMakeLists.txt
+++ b/test/unit/argument_parser/detail/CMakeLists.txt
@@ -1,4 +1,7 @@
-add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_INCLUDE_DIR}")
+# seqan3_test_unit_BINARY_DIR is needed for debian builds
+find_path (SEQAN3_TEST_LICENSE_DIR NAMES LICENSE.md HINTS "${SEQAN3_ROOT_DIR}" "${seqan3_test_unit_BINARY_DIR}/..")
+
+add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_TEST_LICENSE_DIR}")
 seqan3_test(format_help_test.cpp)
 seqan3_test(format_html_test.cpp)
 seqan3_test(format_man_test.cpp)
```

Comment (marehr) [reply test/unit/argument_parser/detail/CMakeLists.txt]: I still don't understand the file layout of debian.

File: test/seqan3-test.cmake [hunk 7]
```diff
@@ -40,7 +40,12 @@ file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
 add_library (seqan3_test INTERFACE)
 target_compile_options (seqan3_test INTERFACE "-pedantic"  "-Wall" "-Wextra" "-Werror")
 target_link_libraries (seqan3_test INTERFACE "seqan3::seqan3" "pthread")
-target_include_directories (seqan3_test INTERFACE "${SEQAN3_CLONE_DIR}/test/include/")
```
Comment (marehr) [inline test/seqan3-test.cmake]: Can you give this patch a try?

```patch
diff --git a/test/seqan3-test.cmake b/test/seqan3-test.cmake
index 0433100a3..e2293d6dd 100644
--- a/test/seqan3-test.cmake
+++ b/test/seqan3-test.cmake
@@ -23,6 +23,8 @@ include (FindPackageMessage)
 # Paths to folders.
 # ----------------------------------------------------------------------------
 
+find_path (SEQAN3_TEST_INCLUDE_DIR NAMES seqan3/test/tmp_filename.hpp HINTS "${CMAKE_CURRENT_LIST_DIR}/include/")
+
 set (SEQAN3_BENCHMARK_CLONE_DIR "${PROJECT_BINARY_DIR}/vendor/benchmark")
 set (SEQAN3_TEST_CLONE_DIR "${PROJECT_BINARY_DIR}/vendor/googletest")
 
@@ -40,7 +42,7 @@ file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
 add_library (seqan3_test INTERFACE)
 target_compile_options (seqan3_test INTERFACE "-pedantic"  "-Wall" "-Wextra" "-Werror" "-DNO_WARN_X86_INTRINSICS")
 target_link_libraries (seqan3_test INTERFACE "seqan3::seqan3" "pthread")
-target_include_directories (seqan3_test INTERFACE "${SEQAN3_CLONE_DIR}/test/include/")
+target_include_directories (seqan3_test INTERFACE "${SEQAN3_TEST_INCLUDE_DIR}")
 add_library (seqan3::test ALIAS seqan3_test)
 
 # seqan3::test::performance specifies required flags, includes and libraries

```

File: test/unit/argument_parser/detail/CMakeLists.txt (hunk 6 above)
Comment (mr-c) [reply test/unit/argument_parser/detail/CMakeLists.txt]: When Seqan3 is installed, `SEQAN_INCLUDE_DIR` is `/usr/include/seqan3/` or `/usr/include` (I forget which). Either way, adding on "../LICENSE.md" won't work

File: test/seqan3-test.cmake (hunk 7 above)
Comment (mr-c) [reply test/seqan3-test.cmake]: both patches or just this one? You should have write permissions to my branch, feel free to commit directly.

File: test/unit/argument_parser/detail/CMakeLists.txt (hunk 6 above)
Comment (marehr) [reply test/unit/argument_parser/detail/CMakeLists.txt]: Okay, but where is `LICENSE.md` located?

File: test/seqan3-test.cmake (hunk 7 above)
Comment (marehr) [reply test/seqan3-test.cmake]: I pushed some changes, can you have a look if they work for you?

**Comment by mr-c** [comment]
@marehr This works for Debian, thanks!

**Comment by mr-c** [comment]
This is ready for (re-)review & merging!

File: test/unit/argument_parser/detail/CMakeLists.txt [hunk 8]
```diff
@@ -1,4 +1,7 @@
-add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_INCLUDE_DIR}")
+# seqan3_test_unit_BINARY_DIR is needed for debian builds
```
Comment (marehr) [inline test/unit/argument_parser/detail/CMakeLists.txt]:
```suggestion

```
remove this line

File: test/unit/argument_parser/detail/CMakeLists.txt [hunk 9]
```diff
@@ -1,4 +1,7 @@
...
+find_path (SEQAN3_TEST_LICENSE_DIR NAMES LICENSE.md HINTS "${SEQAN3_CLONE_DIR}" "${seqan3_test_unit_BINARY_DIR}/..")
```
Comment (marehr) [inline test/unit/argument_parser/detail/CMakeLists.txt]: I want to remove hard-coded paths. 

```suggestion
find_path (SEQAN3_TEST_LICENSE_DIR NAMES LICENSE.md HINTS "${SEQAN3_CLONE_DIR}")
```

You can either add `cmake -DCMAKE_INCLUDE_PATH=..` to influence `find_path` to find `LICENSE.md`, or you can just copy then `LICENSE.md` into the ` test/unit/argument_parser/detail/` directory as a pre-processing step.

See https://cmake.org/cmake/help/v3.16/variable/CMAKE_INCLUDE_PATH.html#variable:CMAKE_INCLUDE_PATH

**Pull request merged**